### PR TITLE
PEM: Add a more generic way to implement PEM _ex functions for libctx

### LINKED ISSUE #13542: Any call to OSSL_ENCODER_CTX_new_byXXX needs to support passing a libctx.
This is related to some new encoder code that ignores the libctx:

e.g:
PEM_write_bio_Parameters() uses the macro IMPLEMENT_PEM_provided_write_body_vars()
which then calls
OSSL_ENCODER_CTX_new_by_##type(x, PEM_SELECTION_##asn1, "PEM", PEM_STRUCTURE_##asn1,  NULL, NULL);
The last 2 params are the libctx, propq which will not work if they are just NULL..

So we either need to extract the libctx (which doesnt address that there is no propq)..
using 
````
const OSSL_PROVIDER *prov = EVP_KEYMGMT_provider(pkey->keymgmt);
OSSL_LIB_CTX *libctx = ossl_provider_libctx(prov);
````
OR have a PEM_write_bio_Parameters_ex(libctx, propq) ??

To reproduce this locally do..
````
export OPENSSL_TEST_LIBCTX=1
apps/openssl genpkey -genparam -algorithm DSA
````

### PR DESCRIPTION
**Pull request by levitte**
This also adds the following functions, for completeness:

PEM_write_PrivateKey_ex(), PEM_write_bio_PrivateKey_ex(),
PEM_write_PUBKEY_ex, PEM_write_bio_PUBKEY_ex

Fixes #13542

### CODE REVIEW

**Comment by levitte** [comment]
This is marked WIP because it may need some mods, depending on what is decided re #13544.

File: crypto/pem/pem_local.h [hunk 1]
```diff
@@ -95,14 +95,22 @@
 # define IMPLEMENT_PEM_provided_write_body_fallback_cb(str, asn1,       \
                                                        writename)       \
     legacy:                                                             \
-    return PEM_ASN1_##writename((i2d_of_void *)i2d_##asn1, str, out,    \
-                                x, enc, kstr, klen, cb, u)
+    return PEM_ASN1_##writename##((i2d_of_void *)i2d_##asn1, str, out,  \
+                                  x, enc, kstr, klen, cb, u)
 
 # define IMPLEMENT_PEM_provided_write_to(name, type, str, asn1,         \
```
Comment (slontis) [inline crypto/pem/pem_local.h:101]: There is no propq in the macro params?

Comment (levitte) [reply crypto/pem/pem_local.h:101]: Why should there be?  `IMPLEMENT_PEM_provided_write_body_vars()` implements the two variables `libctx` and `propq`, which the rest of the building blocks can use.  Why should the implementing macros need to specify those variables explicitly?

Comment (slontis) [reply crypto/pem/pem_local.h:101]: Just a personal style thing I guess.

File: doc/man3/PEM_read_bio_PrivateKey.pod [hunk 2]
```diff
@@ -78,7 +91,11 @@ PEM_write_bio_PKCS7, PEM_write_PKCS7 - PEM routines
                               OSSL_LIB_CTX *libctx, const char *propq);
  EVP_PKEY *PEM_read_PUBKEY(FILE *fp, EVP_PKEY **x,
                            pem_password_cb *cb, void *u);
+ int PEM_write_bio_PUBKEY_ex(BIO *bp, EVP_PKEY *x,
```
Comment (slontis) [inline doc/man3/PEM_read_bio_PrivateKey.pod:94]: Maybe describe the libctx, propq args in the
 PEM FUNCTION ARGUMENTS section?

File: crypto/pem/pem_local.h (hunk 1 above)
Comment (levitte) [reply crypto/pem/pem_local.h:101]: That seems unnecessary, and potentially confusing...

Comment (slontis) [reply crypto/pem/pem_local.h:101]: That is your opinion.

File: doc/man3/PEM_read_bio_PrivateKey.pod (hunk 2 above)
Comment (levitte) [reply doc/man3/PEM_read_bio_PrivateKey.pod:94]: There already is a paragraph at the end of that section.

**Comment by levitte** [comment]
This is not pending on anything any more

**Comment by levitte** [comment]
CIs are happy

File: crypto/pem/pem_local.h [hunk 3]
```diff
@@ -113,11 +121,19 @@
                                             OUTTYPE, outtype, writename) \
     PEM_write_cb_fnsig(name, type, OUTTYPE, writename)                  \
     {                                                                   \
-        IMPLEMENT_PEM_provided_write_body_vars(type, asn1);             \
+        IMPLEMENT_PEM_provided_write_body_vars(type, asn1, NULL);       \
```
Comment (slontis) [inline crypto/pem/pem_local.h:124]: Where is the libctx? It is in the macro def but not here.

File: crypto/pem/pem_local.h [hunk 4]
```diff
@@ -113,11 +121,19 @@
...
         IMPLEMENT_PEM_provided_write_body_pass();                       \
         IMPLEMENT_PEM_provided_write_body_main(type, outtype);          \
         IMPLEMENT_PEM_provided_write_body_fallback_cb(str, asn1,        \
                                                       writename);       \
+    }                                                                   \
+    PEM_write_ex_cb_fnsig(name, type, OUTTYPE, writename)               \
+    {                                                                   \
+        IMPLEMENT_PEM_provided_write_body_vars(type, asn1, propq);      \
```
Comment (slontis) [inline crypto/pem/pem_local.h:132]: Same here..

File: crypto/pem/pem_pkey.c [hunk 5]
```diff
@@ -212,8 +218,7 @@ EVP_PKEY *PEM_read_bio_Parameters(BIO *bp, EVP_PKEY **x)
 PEM_write_fnsig(Parameters, EVP_PKEY, BIO, write_bio)
 {
     char pem_str[80];
-    IMPLEMENT_PEM_provided_write_body_vars(EVP_PKEY, Parameters);
-
+    IMPLEMENT_PEM_provided_write_body_vars(EVP_PKEY, Parameters, NULL, NULL);
```
Comment (slontis) [inline crypto/pem/pem_pkey.c:221]: blankline

File: crypto/pem/pem_local.h (hunk 3 above)
Comment (levitte) [reply crypto/pem/pem_local.h:124]: Ah, I apparently didn't fix up everything

**Comment by levitte** [comment]
Ping

File: doc/man3/PEM_read_bio_PrivateKey.pod (hunk 2 above)
Comment (slontis) [reply doc/man3/PEM_read_bio_PrivateKey.pod:94]: Yeah it is about 2 pages of content down :)

**Review by slontis — APPROVED** [review]

**Comment by levitte** [comment]
Merged

9256e8a2487697c347f8e54db69509dc662b26f7 PEM: Add a more generic way to implement PEM _ex functions for libctx

**Pull request closed without merging**